zend_abstract_interface/zai_string/string.h

```diff
@@ -11,25 +11,46 @@
  * Represents a non-owning view of a string.
  *
  * When initializing the struct, use one of the initialization macros or
- * functions. Do not initialize the struct directly e.g. `{0, null}`.
- *
- * todo: move .ptr to come before .len to match ddtrace_string and Rust and
- *       ensure it's never null.
+ * functions. Do not initialize the struct directly e.g. `{"", 0}`.
  */
 typedef struct zai_str_s {
-    size_t len;
     const char *ptr;
+    size_t len;
 } zai_str;
 
-/** Use if data is known to be non-null, use zai_str_new otherwise. */
-#define ZAI_STR_NEW(data, size)   \
-    (zai_str) {.len = (size), .ptr = (data)}
+/**
+ * ZAI_ASSERT is like ZEND_ASSERT and C assert that it will expand into a valid
+ * expression which returns true (if it fails, it will not return at all).
+ */
+#if NDEBUG
+#define ZAI_ASSERT(cond) true
+#else
+#define ZAI_ASSERT(cond) (assert(cond), true)
```

Comment:
Should I maybe move this somewhere else, or rename it to be ZA_STR_ASSERT or something?